Infrastructure/Services/ProcessEmailsService.cs

```diff
@@ -0,0 +1,482 @@
+
+namespace Infrastructure.Services;
+
+public class ProcessEmailsService : IProcessEmailsService
+{
+    private const int ProcessingStillRunningForMinutesThrowError = 30;
+    private const int ProcessingStillRunningForMinutesLogWarning = 5;
+    private const string MergedEmailAddress = "merged_email_for_testing@education.gov.uk";
+    private const string ScheduleName = "Process Emails";
+    private const string TestExtraInfoKey = "test_extra_info";
+
+    private readonly IUnitOfWork _unitOfWork;
+    private readonly ILogger<ProcessEmailsService> _logger;
+    private readonly INotificationsClientService _notificationsClientService;
+    private readonly EmailSettings _emailSettingsConfig;
+
+    public ProcessEmailsService(IUnitOfWork unitOfWork,
+        INotificationsClientService notificationsClientService,
+        ILogger<ProcessEmailsService> logger,
+        IOptions<EmailSettings> emailSettingsConfig)
+    {
+        _unitOfWork = unitOfWork;
+        _notificationsClientService = notificationsClientService;
+        _logger = logger;
+        _emailSettingsConfig = emailSettingsConfig.Value;
+    }
+
+    public async Task<ProcessedEmailsModel> ProcessAllEmailsAsync()
+    {
+        var processedEmailsModel = new ProcessedEmailsModel();
+        var startDate = DateTime.UtcNow;
+        _logger.LogInformation("Starting ProcessAllEmailsAsync at {startDate}", startDate);
+
+        try
+        {
+            var startedOK = await StartProcessing(startDate);
+            if (startedOK)
+            {
+                if (_emailSettingsConfig.SendEmailsFromNtp)
+                {
+                    await PollForStatusUpdatesAsync(processedEmailsModel);
+
+                    processedEmailsModel.EmailsSent = await SendPendingEmailsAsync(null, false);
+
+                    var finishedDate = DateTime.UtcNow;
+                    var finishedMessage = $"Completed email process successfully.  Started at {startDate}, finished at {finishedDate}, taking {(finishedDate - startDate).TotalSeconds} seconds.  " +
+                        $"{processedEmailsModel.EmailsCheckStatus} email status checks, {processedEmailsModel.EmailsUpdatedStatus} changed email status, {processedEmailsModel.EmailsSent} emails sent.";
+
+                    await FinishProcessing(finishedDate, finishedMessage);
+
+                    processedEmailsModel.Outcome = finishedMessage;
+                }
+                else
+                {
+                    await DeactivateEmailsAsync();
+                    processedEmailsModel.Outcome = "Sending of emails is switched off in the config";
+                    await FinishProcessing(DateTime.UtcNow, processedEmailsModel.Outcome);
+                }
+            }
+            else
+            {
+                processedEmailsModel.Outcome = $"Previous email process is still running at {startDate}.";
+            }
+
+            _logger.LogInformation(processedEmailsModel.Outcome);
```

Comment:
@kieranohanlon Can we look into that?

Reply:
I've already looked in to this and added a comment to the scanning alert that I don't believe this is logging any sensitive data, it's logging exactly the same outcome as we display on the polling web page.  But if you think otherwise let me know.